### PR TITLE
feat: [CDS-78341] Move K8s BG Step service to be AsyncExecutable

### PR DESCRIPTION
**Pull request by alebuh**
## Describe your changes
Changes explained in the tech spec https://harness.atlassian.net/wiki/spaces/CDP/pages/21515076842/CD+Docs+on+moving+to+Async+Step

### Test cases
- [x] 1. When FF is disabled use sync (V1) step implementation
- [x] 2. When FF is enabled use async (V2) step implementation
Screenshot 2023-10-11 at 15 39 23

 - Execution was successful
 - The icon should be mapped to the new step type on the UI
 - Logs are not present during the execution, only at the end of it
 - Logs for delegate selector is present
 - Step outputs are populated correctly

- [x] 3. Step execution failure - Exceptions have been handled by the Error framework
- [x] 4. Step execution abortion - If the Step is aborted, the task should be aborted on the delegate side as well
 - Logs are missing if the step is aborted

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53905)

### CODE REVIEW

**Comment by alebuh** [comment]
trigger uts

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53905)* status: 10 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): del-team,k8s-team. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/executables/CdAsyncChainExecutable.java` line 44 at r1](https://reviewable.io/reviews/harness/harness-core/53905#-NgNGe6E6LLNlwU9NX-R:-NgNGe6E6LLNlwU9NX-S:brervc5) ([raw file](https://github.com/harness/harness-core/blob/34085de0c59134e9da03737ead0d050aa3d15f76/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/executables/CdAsyncChainExecutable.java#L44)):*
> ```Java
> public class CdAsyncChainExecutable<T extends CdTaskChainExecutable>
>     implements AsyncChainExecutableWithRbac<StepBaseParameters> {
>   @Inject private T cdTaskChainExecutable;
> ```

instead of we injecting this, can we make abstract class of say bluegreenstep step, and both task and async are extending it? this will have clear boundaries
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/executables/CdAsyncExecutable.java` line 35 at r1](https://reviewable.io/reviews/harness/harness-core/53905#-NgNG7F-6u8nLOAEMHZC:-NgNG7F-6u8nLOAEMHZD:b-7sho66) ([raw file](https://github.com/harness/harness-core/blob/34085de0c59134e9da03737ead0d050aa3d15f76/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/executables/CdAsyncExecutable.java#L35)):*
> ```Java
> @OwnedBy(CDC)
> @Slf4j
> public class CdAsyncExecutable<T extends CdTaskExecutable<ResponseData>> extends AsyncExecutableWithCapabilities {
> ```

why async executable should extend taskExecutable?



_Code quote:_
```Java
CdTaskExecutable
```

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger uts

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 1]
```diff
@@ -931,7 +931,8 @@ public enum FeatureName {
       "This flag to not support secrets in files which content will be used in configFile.getAsBase64 and fileStore.getAsBase64 expressions",
       HarnessTeam.CDP),
   CDS_INFRA_GITX("Onboard Infrastructure entity to GitX", HarnessTeam.CDC),
-  PIE_SECRETS_OBSERVER("Secrets Runtime Usages Observer", PIPELINE);
+  PIE_SECRETS_OBSERVER("Secrets Runtime Usages Observer", PIPELINE),
+  CDS_K8S_ASYNC_STEP_STRATEGY("This Flag enables the usage of ASYNC chain orchestration for K8s step", HarnessTeam.CDP);
```
Comment (TarunUba1) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:935]: we might have to confirm on this FF with rohan

**Review by TarunUba1 — DISMISSED** [review]

**Comment by alebuh** [comment]
trigger ut9

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger uts

**Review by abosii-harness — APPROVED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/asyncsteps/K8sBlueGreenStepV2.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import io.harness.annotations.dev.ProductModule;
+import io.harness.cdng.executables.CdAsyncChainExecutable;
+import io.harness.cdng.k8s.K8sBlueGreenStep;
+import io.harness.executions.steps.ExecutionNodeType;
+import io.harness.pms.contracts.steps.StepCategory;
+import io.harness.pms.contracts.steps.StepType;
+
+import lombok.extern.slf4j.Slf4j;
+
+@CodePulse(module = ProductModule.CDS, unitCoverageRequired = true, components = {HarnessModuleComponent.CDS_K8S})
```
Comment (abhinavsingh1196) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/asyncsteps/K8sBlueGreenStepV2.java:23]: nit: unitCoverageRequired = false?

**Review by abhinavsingh1196 — APPROVED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/asyncsteps/K8sBlueGreenStepV2.java (hunk 2 above)
Comment (alebuh) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/k8s/asyncsteps/K8sBlueGreenStepV2.java:23]: yes, will update it in the next PR

**Review by yogesh-chauhan — APPROVED** [review]

**Review by archit-harness — COMMENTED** [review]
> Reviewed 3 of 10 files at r1, 2 of 9 files at r2, 1 of 1 files at r3, 5 of 8 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53905)* status: 14 file(s) to review. 1 change requests. 1. unresolved discussions.

**Review by archit-harness — APPROVED** [review]
> Reviewed 2 of 8 files at r4.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53905)* status: 14 file(s) to review. 1. unresolved discussions.

**Comment by alebuh** [comment]
trigger review

**Review by jennyjames21 — APPROVED** [review]

**Pull request merged**